runtime/server/x86/frontend/fbank.h

```diff
@@ -0,0 +1,212 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef FRONTEND_FBANK_H_
+#define FRONTEND_FBANK_H_
+
+#include <assert.h>
+
+#include <vector>
+#include <limits>
+#include <random>
+#include <utility>
+
+#include "frontend/fft.h"
+
+#ifndef M_PI
+#define M_PI 3.1415926535897932384626433832795
+#endif
+#ifndef M_2PI
+#define M_2PI 6.283185307179586476925286766559005
+#endif
+
+namespace wenet {
+
+class Fbank {
+ public:
+  Fbank(int num_bins, int sample_rate, int frame_length, int frame_shift):
+      num_bins_(num_bins), sample_rate_(sample_rate),
+      frame_length_(frame_length), frame_shift_(frame_shift),
+      use_log_(true), remove_dc_offset_(true),
+      generator_(0), distribution_(0, 1.0), dither_(0.0) {
+    fft_points_ = UpperPowerOfTwo(frame_length_);
+    int num_fft_bins = fft_points_ / 2;
+    float fft_bin_width = static_cast<float>(sample_rate_) / fft_points_;
+    int low_freq = 20, high_freq = sample_rate_ / 2;
+    float mel_low_freq = MelScale(low_freq);
+    float mel_high_freq = MelScale(high_freq);
+    float mel_freq_delta = (mel_high_freq - mel_low_freq) / (num_bins+1);
+    bins_.resize(num_bins_);
+    center_freqs_.resize(num_bins_);
+    for (int bin = 0; bin < num_bins; bin++) {
+      float left_mel = mel_low_freq + bin * mel_freq_delta,
+      center_mel = mel_low_freq + (bin + 1) * mel_freq_delta,
+      right_mel = mel_low_freq + (bin + 2) * mel_freq_delta;
+      center_freqs_[bin] = InverseMelScale(center_mel);
+      std::vector<float> this_bin(num_fft_bins);
+      int first_index = -1, last_index = -1;
+      for (int i = 0; i < num_fft_bins; i++) {
+        float freq = (fft_bin_width * i);  // Center frequency of this fft
+        // bin.
+        float mel = MelScale(freq);
+        if (mel > left_mel && mel < right_mel) {
+          float weight;
+          if (mel <= center_mel)
+            weight = (mel - left_mel) / (center_mel - left_mel);
+          else
+            weight = (right_mel-mel) / (right_mel-center_mel);
+          this_bin[i] = weight;
+          if (first_index == -1)
+            first_index = i;
+          last_index = i;
+        }
+      }
+      assert(first_index != -1 && last_index >= first_index);
+      bins_[bin].first = first_index;
+      int size = last_index + 1 - first_index;
+      bins_[bin].second.resize(size);
+      for (int i = 0; i < size; i++) {
+        bins_[bin].second[i] = this_bin[first_index + i];
+      }
+    }
+
+    // hamming window
+    hamming_window_.resize(frame_length_);
+    double a = M_2PI / (frame_length-1);
+    for (int i = 0; i < frame_length; i++) {
+        hamming_window_[i] = 0.54 - 0.46*cos(a * i);
+    }
+  }
+
+  void SetUseLog(bool use_log) {
```

Comment:
> we have a member named input_finished


Why not use `set_use_log`?  Do you have any coding guidelines to follow?

Reply:
fixed